### PR TITLE
immediately close the other side of the connection when proxying

### LINKED ISSUE #57992: Kubelet leaks memory on kubectl port-forward
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
When opening port-forward requests to a pod, kubelet memory usage increases without releasing it again when the port-forward is stopped.

**What you expected to happen**:
Kubelet should cleanup the port-forward when it is closed

**How to reproduce it (as minimally and precisely as possible)**:
Using the following script I can observe an increase in Kubelet memory usage on the node where the pod runs:

```bash
for i in {8000..8300}; do
kubectl port-forward anypod $i:8080 &; sleep 0.2;
done

...
killall kubectl
```
300 connections allocates roughly 30mb by kubelet

We first noticed this on the node where Tiller (Helm) runs. The node regularly runs out of memory because of the Kubelet memory usage. see also https://github.com/kubernetes/helm/issues/3121 We don't experience the helm issue on all clusters, but I can reproduce the memory leak with the above script on all clusters.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"clean", BuildDate:"2018-01-04T19:59:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.11", GitCommit:"b13f2fd682d56eab7a6a2b5a1cab1a3d2c8bdd55", GitTreeState:"clean", BuildDate:"2017-11-25T17:51:39Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): Debian jessie
- Kernel (e.g. `uname -a`): `Linux 4.4.102-k8s #1 SMP Sun Nov 26 23:32:43 UTC 2017 x86_64 GNU/Linux`
- Install tools: Kops 1.8.0
- Others:

### PR DESCRIPTION
**Pull request by MHBauer**
**What this PR does / why we need it**:
close the other side of the connection by exiting once one side closes. This code is used in both apiserver and kubelet.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #57992

**Special notes for your reviewer**:
unit tests passed locally. something similar is done in client-go/tools/portforward. doesn't seem to break the tests.

**Release note**:
```release-note
Immediately close the other side of the connection when proxying.
```

### CODE REVIEW

**Comment by duglin** [comment]
LGTM

**Comment by MHBauer** [comment]
FYI @mikebrow @Random-Liu Not sure if there is a better node based fix from CRI pushing events to close upwards. Thanks for talking through this with me.
/sig api-machinery

**Review by mikebrow — COMMENTED** [review]
NIce!

nit.. since we have log for creating the connection, maybe we could use log on closing the connection in the defer?

**Comment by MHBauer** [comment]
@mikebrow before and after the wait? One side or the other specifically?

**Comment by mikebrow** [comment]
Well yes, but I was specifically looking at these "Connecting to backend proxy"  entires: https://github.com/kubernetes/kubernetes/pull/67288/files?utf8=✓&diff=unified#diff-16a26bb0d342318c6eb3d03e6c5756e9R257

and thinking we should have a reciprocal closing message.

**Comment by caesarxuchao** [comment]
I have a stupid question, from https://github.com/kubernetes/kubernetes/issues/57992#issuecomment-409740452,

> After closing a port-forward from kubectl, I see immediately the reader close on the apiserver, and the writer close on the kubelet.

Why is that? Shouldn't both the reader and writer close on both apiserver and kubelet?

**Comment by MHBauer** [comment]
@mikebrow I put one extra log message in, with a similar format to the opening connection log message you pointed at. Thanks.

**Comment by anguslees** [comment]
So ... Let's say I have this exchange:
- client send request
- client closes the write direction (ie: shutdown() syscall, ie: not http-keepalive)
- server does some non-trivial processing
- server sends reply

Will the connection still be open for the reply?  I haven't built it and tried yet, but I gather from the description and the patch that this will kill both directions of the connection when either closes - breaking the above use of not-HTTP-keepalive.  Is that correct?

Fwiw, in an ssh / `kubectl exec` context, the common equivalent of the above scenario is piping to `sort` on the remote side.

**Comment by caesarxuchao** [comment]
@MHBauer I thought the io.Copy would only return after after the request connection or the backend connection is closed, in that case, the PR was right. But io.Copy returns at EOF, so @anguslees's comment makes sense.

**Comment by MHBauer** [comment]
@anguslees 
not sure what what the case is.

sort pipe like this? Something else?
```
     Running `genrand 10000000`
➜  ls -lh
total 195M
-rw-rw-r-- 1 mhb mhb 195M Aug 17 09:40 nums
➜  sort nums | wc -l
10000000
➜  sort nums | wc -l
➜  docker run -i -a stdin -a stdout busybox sort - < nums | wc -l
10000000
➜  kubectl get nodes                                             
NAME          STATUS    ROLES     AGE       VERSION
kube-master   Ready     master    15m       v1.13.0-alpha.0.234+b9544382baf4a0-dirty
kube-node-1   Ready     <none>    14m       v1.13.0-alpha.0.234+b9544382baf4a0-dirty
kube-node-2   Ready     <none>    14m       v1.13.0-alpha.0.234+b9544382baf4a0-dirty
➜  kubectl version
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.1", GitCommit:"b1b29978270dc22fecc592ac55d903350454310a", GitTreeState:"clean", BuildDate:"2018-08-07T04:27:44Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.0-alpha.0.234+b9544382baf4a0-dirty", GitCommit:"b9544382baf4a0a366dd01e05e585f8556993754", GitTreeState:"dirty", BuildDate:"2018-08-17T16:23:52Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

➜  time kubectl run bb --image busybox --restart=Never -i --rm sort - < nums | wc -l
If you don't see a command prompt, try pressing enter.
10000001
kubectl run bb --image busybox --restart=Never -i --rm sort - < nums  1.33s user 1.23s system 3% cpu 1:11.84 total
wc -l  0.13s user 0.16s system 0% cpu 1:11.84 total
➜  
     Running `genrand 10`
➜  cat nums 
196211602033574003
5523453742303018414
15316494326757210401
5642514980812386917
6611172744505202147
17001739917637352584
5020503333694379813
9999606963224618426
18295755340485735596
8929362997318480641
➜  time sort nums                                                
15316494326757210401
17001739917637352584
18295755340485735596
196211602033574003
5020503333694379813
5523453742303018414
5642514980812386917
6611172744505202147
8929362997318480641
9999606963224618426
sort nums  0.00s user 0.00s system 83% cpu 0.003 total
➜  docker run -i -a stdin -a stdout busybox sort - < nums        
15316494326757210401
17001739917637352584
18295755340485735596
196211602033574003
5020503333694379813
5523453742303018414
5642514980812386917
6611172744505202147
8929362997318480641
9999606963224618426
➜  kubectl run bb --image busybox --restart=Never -i --rm sort - < nums        
If you don't see a command prompt, try pressing enter.
15316494326757210401
17001739917637352584
18295755340485735596
196211602033574003
5020503333694379813
5523453742303018414
5642514980812386917
6611172744505202147
8929362997318480641
9999606963224618426
pod "bb" deleted
```

**Comment by MHBauer** [comment]
just re-read and saw the exec. that also works with the busybox pod running with an open sh in another terminal.
```
➜  kubectl exec -i bb  sort - < nums                                   
15316494326757210401
17001739917637352584
18295755340485735596
196211602033574003
5020503333694379813
5523453742303018414
5642514980812386917
6611172744505202147
8929362997318480641
9999606963224618426
➜
     Running `genrand 10000000`
     
➜  kubectl exec -i bb  sort - < nums | wc -l
10000000
```

**Comment by MHBauer** [comment]
Do you have a test case? I'm not sure I fully understand.

**Comment by anguslees** [comment]
> Do you have a test case?

Sorry for the delay, it took me a while to find a good test client that actually made a TLS HTTP/1.0 request, with half-close (it appears curl can't be made to do this).

I constructed a full test scenario that generated an HTTP/1.0 TLS request with an early close to a /api/.../proxy' URL for a job that introduced a delay before responding ... and found that my nearby k8s 1.8 (old!) test cluster already "failed" my test.   Then I went back and re-read the HTTP/1.0 spec and found that this was acceptable behaviour ("the closing of the connection by either or both parties always terminates the current request, regardless of its status.")

So: My concern about doing this at the HTTP-level was invalid, and your testing above confirms that there's no problem at the tunneled-over-websocket level.

TL;DR: lgtm!  And sorry for the distraction.

**Comment by mbohlool** [comment]
@MHBauer I added a release note to this PR because we need one for the cherrypick. please see if that looks good.

**Comment by MHBauer** [comment]
@anguslees thanks for the followup.

@mbohlool release note looks fine. Sorry for my hold up in responding.

**Comment by MHBauer** [comment]
Reading the cherry-pick discussions, this isn't a user facing change. There should be nothing different. It changes how server components interact with each other.
